File: rust_marcenaria_actix/src/models/geometrias/geometria.rs
use std::f64::consts::PI;

use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::utils::str_utils;

#[derive(Serialize, Deserialize, ToSchema)]
#[serde(tag = "geometria")]
pub enum Geometria {
    #[serde(alias = "esfera", alias = "Esfera", alias = "ESFERA")]
    Esfera {
        #[serde(skip_serializing)]
        estrutura: String,
        raio: String,
    },
    #[serde(alias = "cubo", alias = "Cubo", alias = "CUBO")]
    Cubo {
        #[serde(skip_serializing)]
        estrutura: String,
        lado: String,
    },
    #[serde(alias = "cilindro", alias = "Cilindro", alias = "CILINDRO")]
    Cilindro {
        #[serde(skip_serializing)]
        estrutura: String,
        raio_base: String,
        altura: String,
    },
    #[serde(
        alias = "paralelepipedo",
        alias = "paralelepípedo",
        alias = "Paralelepípedo",
        alias = "Paralelepipedo",
        alias = "PARALELEPIPEDO",
        alias = "PARALELEPÍPEDO"
    )]
    Paralelepipedo {
        #[serde(skip_serializing)]
        estrutura: String,
        comprimento: String,
        largura: String,
        altura: String,
    },
    Desconhecido,
}

impl Geometria {
    pub fn get_all_info() -> [Geometria; 4] {
        [
            Geometria::Esfera {
                estrutura: String::default(),
                raio: String::from("1,00cm"),
            },
            Geometria::Cubo {
                estrutura: String::default(),
                lado: String::from("1,00cm"),
            },
            Geometria::Cilindro {
                estrutura: String::default(),
                raio_base: String::from("1,00cm"),
                altura: String::from("1,00cm"),
            },
            Geometria::Paralelepipedo {
                estrutura: String::default(),
                comprimento: String::from("1,00cm"),
                largura: String::from("1,00cm"),
                altura: String::from("1,00cm"),
            },
        ]
    }
}

impl Geometria {
    pub fn get_estrutura(&self) -> Option<&str> {
        match self {
            Geometria::Esfera { estrutura, .. } => Some(estrutura),
            Geometria::Cubo { estrutura, .. } => Some(estrutura),
            Geometria::Cilindro { estrutura, .. } => Some(estrutura),
            Geometria::Paralelepipedo { estrutura, .. } => Some(estrutura),
            Geometria::Desconhecido => None,
        }
    }

    pub fn get_name(&self) -> Result<&'static str, &'static str> {
        match self {
            Geometria::Esfera {..} => Ok("Esfera"),
            Geometria::Cubo {..} => Ok("Cubo"),
            Geometria::Cilindro {..} => Ok("Cilindro"),
            Geometria::Paralelepipedo {..} => Ok("Paralelepípedo"),
            _ => Err("Geometria inválida, por favor, utilize apenas geometrias válidas do endpoint /geometrias")
        }
    }

    pub fn get_area(&self) -> Result<f64, String> {
        match self {
            Geometria::Esfera { raio, estrutura } => {
                let r: f64 = str_utils::convert_to_f64(raio).map_err(|_| {
                    format!(
                        "A esfera {} possui o valor do raio inválido: {}",
                        estrutura, raio
                    )
                })?;

                if r <= 0.0 {
                    let err_message = format!(
                        "A esfera {} deve ter um raio positivo, {} é inválido",
                        estrutura, raio
                    );
                    Err(err_message)
                } else {
                    Ok(4.0 * PI * r * r)
                }
            }

            Geometria::Cubo { estrutura, lado } => {
                let a = str_utils::convert_to_f64(lado).map_err(|_| {
                    format!(
                        "O cubo {} está com o valor do lado inválido: {}",
                        estrutura, lado
                    )
                })?;

                if a <= 0.0 {
                    let err_message = format!(
                        "O cubo {} deve ter o lado positivo, {} é inválido",
                        estrutura, lado
                    );
                    Err(err_message)
                } else {
                    Ok(6.0 * a * a)
                }
            }

            Geometria::Cilindro {
                estrutura,
                raio_base,
                altura,
            } => {
                let r = str_utils::convert_to_f64(raio_base).map_err(|_| {
                    format!(
                        "O cilindro {} está com o valor do raio inválido {}",
                        estrutura, raio_base
                    )
                })?;

                let a = str_utils::convert_to_f64(altura).map_err(|_| {
                    format!(
                        "O cilindro {} está com o valor da altura inválida: {}",
                        estrutura, altura
                    )
                })?;

                if a <= 0.0 || r <= 0.0 {
                    let err_message = format!(
            "O cilindro {} deve ter a altura e o raio_base positivos, raio: {} ou altura: {} é inválido",
            estrutura, raio_base, altura
          );
                    Err(err_message)
                } else {
                    Ok(2.0 * PI * r * (r + a))
                }
            }

            Geometria::Paralelepipedo {
                estrutura,
                comprimento,
                largura,
                altura,
            } => {
                let c = str_utils::convert_to_f64(comprimento).map_err(|_| {
                    format!(
                        "O paralelepípedo {} está com o valor do comprimento inválido: {}",
                        *estrutura, *comprimento
                    )
                })?;

                let l = str_utils::convert_to_f64(largura).map_err(|_| {
                    format!(
                        "O paralelepípedo {} está com o valor da largura inválido: {}",
                        *estrutura, *largura
                    )
                })?;

                let a = str_utils::convert_to_f64(altura).map_err(|_| {
                    format!(
                        "O paralelepípedo {} está com o valor da altura inválido: {}",
                        *estrutura, *altura
                    )
                })?;

                if c <= 0.0 || l <= 0.0 || a <= 0.0 {
                    let err_message = format!(
            "O paralelepípedo {} deve ter o comprimento, largura e altura positivos, comprimento: {}, altura: {} ou largura {} é inválido",
            *estrutura, *comprimento, *largura, *altura
        );
                    Err(err_message)
                } else {
                    Ok(2.0 * (c * l + l * a + a * c))
                }
            }

            _ => Err(String::from("err")),
        }
    }
}

File: marcenaria_go_hertz/models/geometria.go
package models

import (
	"errors"
	"fmt"

	"marcenaria_go_hertz/utils"
)

type Geometria struct {
	Estrutura   string `json:"estrutura,omitempty"`
	Geometria   string `json:"geometria"`
	RaioBase    string `json:"raio_base,omitempty"`
	Lado        string `json:"lado,omitempty"`
	Raio        string `json:"raio,omitempty"`
	Altura      string `json:"altura,omitempty"`
	Comprimento string `json:"comprimento,omitempty"`
	Largura     string `json:"largura,omitempty"`
}

const defaultCm = "5,0cm"

var Geometrias = [4]Geometria{
	{
		Geometria: "Esfera",
		Raio:      defaultCm,
	},
	{
		Geometria: "Cubo",
		Lado:      defaultCm,
	},
	{
		Geometria: "Cilindro",
		RaioBase:  defaultCm,
		Altura:    defaultCm,
	},
	{
		Geometria:   "Paralelepipedo",
		Comprimento: defaultCm,
		Largura:     defaultCm,
		Altura:      defaultCm,
	},
}

func (g Geometria) GetPreco(material *Material) (float64, []error) {
	area, errList := g.GetArea()

	if errList != nil {
		return 0, errList
	}

	precoMaterial, err := utils.ExtractDecimal(material.Preco)
	if err != nil {
		errList = append(errList, err)
		return 0, errList
	}

	preco := area * precoMaterial

	return preco, nil
}

func (g Geometria) GetArea() (float64, []error) {
	errList := make([]error, 0)
	geometria := utils.NormalizeStr(g.Geometria)

	var area float64

	switch geometria {
	case "esfera":
		area = g.getAreaEsfera(&errList)
		break
	case "cubo":
		area = g.getAreaCubo(&errList)
		break
	case "cilindro":
		area = g.getAreaCilindro(&errList)
		break
	case "paralelepipedo":
		area = g.getAreaParalelepipedo(&errList)
		break
	default:
		err := fmt.Sprintf("Geometria %s não encontrada", g.Geometria)
		errList = append(errList, errors.New(err))
		break
	}

	if len(errList) > 0 {
		return 0, errList
	}

	return area, nil
}

func (g Geometria) getAreaEsfera(errList *[]error) float64 {
	if g.Raio == "" {
		err := fmt.Sprintf("Raio da esfera %s não informado", g.Estrutura)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	raio, err := utils.ExtractDecimal(g.Raio)

	if err != nil || raio < 0 {
		err := fmt.Sprintf(
			"Raio da esfera %s inválido, deve ser um número positivo e válido",
			g.Estrutura,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	area := 4 * 3.14 * raio * raio

	return area
}

func (g Geometria) getAreaCubo(errList *[]error) float64 {
	if g.Lado == "" {
		err := fmt.Sprintf("Lado do cubo %s não informado", g.Estrutura)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	lado, err := utils.ExtractDecimal(g.Lado)

	if err != nil || lado < 0 {
		err := fmt.Sprintf(
			"Lado do cubo '%s' é inválido: %s, deve ser um número positivo e válido",
			g.Estrutura,
			g.Lado,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	area := 6 * lado * lado

	return area
}

func (g Geometria) getAreaCilindro(errList *[]error) float64 {
	if g.Altura == "" || g.RaioBase == "" {
		err := fmt.Sprintf("Altura ou raio da base do cilindro %s não foi informado", g.Estrutura)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	altura, err := utils.ExtractDecimal(g.Altura)

	if err != nil || altura < 0 {
		err := fmt.Sprintf(
			"Altura do cilindro %s é inválida, deve ser um número positivo e válido",
			g.Estrutura,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	raioBase, err := utils.ExtractDecimal(g.RaioBase)

	if err != nil || raioBase < 0 {
		err := fmt.Sprintf(
			"Raio da base do cilindro %s é inválido, deve ser um número positivo e válido",
			g.Estrutura,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	area := 2 * 3.14 * raioBase * (altura + raioBase)
	return area
}

func (g Geometria) getAreaParalelepipedo(errList *[]error) float64 {
	if g.Comprimento == "" || g.Largura == "" || g.Altura == "" {
		err := fmt.Sprintf(
			"Comprimento, largura ou altura do paralelepipedo %s não foi informado",
			g.Estrutura,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	comprimento, err := utils.ExtractDecimal(g.Comprimento)

	if err != nil || comprimento < 0 {
		err := fmt.Sprintf(
			"Comprimento do paralelepipedo %s é inválido, deve ser um número positivo e válido",
			g.Estrutura,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	largura, err := utils.ExtractDecimal(g.Largura)

	if err != nil || largura < 0 {
		err := fmt.Sprintf(
			"Largura do paralelepipedo %s é inválida, deve ser um número positivo e válido",
			g.Estrutura,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	altura, err := utils.ExtractDecimal(g.Altura)

	if err != nil || altura < 0 {
		err := fmt.Sprintf(
			"Altura do paralelepipedo %s é inválida, deve ser um número positivo e válido",
			g.Estrutura,
		)
		*errList = append(*errList, errors.New(err))
		return 0
	}

	area := 2 * (comprimento*largura + comprimento*altura + largura*altura)

	return area
}
